
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Steps &#8212; bgwas3 0.0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="_static/graphviz.css" />
    <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="steps">
<h1>Steps<a class="headerlink" href="#steps" title="Permalink to this headline">Â¶</a></h1>
<div class="graphviz"><img src="_images/graphviz-aa60efa298b1247050ab0768664d2a9a05c59db0.png" alt="digraph {

	graph [ dpi = 300 ]; 

	node [ 
		fontname=&quot;Sans serif&quot; 
		fontsize=&quot;36&quot;
		shape=folder
		style=filled
		fillcolor=white
	];

	subgraph cluster_input_files { /* {{{ */
		fontname=&quot;Sans serif&quot; 
		fontsize=36
		label = &quot;Input files&quot;;
		contigs [ 
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Sample sequences&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;*.fa&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		pheno [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Table of phenotypes&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		refs[ 
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;External reference files&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.fa, .gff&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		yml [ 
			shape=note
			fillcolor=yellow
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Configuration file&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.yml&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];

	} /* }}} */

	/* once {{{ */
	indexes [ 
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Index&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.amb, .ann, .bwt, .pac, .sa&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	annotations [ 
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Sample annotations&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;gff&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	annotations_bed [ 
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Sample annotations&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.bed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	phenos [ 
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Phenoypes files&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	refs_bed [ 
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Reference annotations&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.bed&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	phylogenetic_tree [ 
		shape=note
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Phylogenetic tree&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.newick&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	bwa_index [ 
		shape=ellipse
		style=filled
		fillcolor=grey
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;bwa_index&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;BWA index&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	mine_kmers [ 
		shape=ellipse
		style=filled
		fillcolor=grey
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;mine_kmers&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;FSM-lite&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	annotate [ 
		shape=ellipse
		style=filled
		fillcolor=grey
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;annotate&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;prokka&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	pangenome_analysis [ 
		shape=ellipse
		style=filled
		fillcolor=grey
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;pangenome_analysis&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;Roary&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	distance_from_tree [ 
		shape=ellipse
		style=filled
		fillcolor=grey
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;distance_from_tree&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;distance_from_tree.py&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	plot_tree [ 
		shape=ellipse
		style=filled
		fillcolor=grey
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;plot_tree&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;plot_tree.R&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	split_phenos [ 
		shape=ellipse
		style=filled
		fillcolor=grey
		label=&lt;
			&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;split_phenos&lt;/td&gt;&lt;/tr&gt;
				&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;split_phenos.R&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
			&lt;/table&gt;
		&gt;
	];
	annotation2bed [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;annotation2bed&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;gff2bed, bedtools merge&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		ref2bed [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;ref2bed&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;gff2bed, bedtools merge&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		make_ref_list [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;make_ref_list&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		make_index [ 
			shape=ellipse
			style=filled

			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;make_index&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
	/* }}} */

	subgraph cluster_rerun { /* {{{ */
		fontname=&quot;Sans serif&quot; 
		fontsize=36
		label=&quot;Run for every phenotype&quot;
		style = filled;
		color = lightblue;
		stats [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Summary statistics&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		kmers [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Kmers&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.txt.gz&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		distance_matrix [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Distance matrix&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
				&gt;
			];
		pheno_i [ 
				shape=note
				label=&lt;
					&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
						&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Phenotype file&lt;/td&gt;&lt;/tr&gt;
						&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
					&lt;/table&gt;
				&gt;
			];
		associated_kmers [ 
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Kmers and statistics of association&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		threshold[ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Threshold&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		kmers_filtered [ 
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Kmers and statistics of association (filtered)&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		kmers_mapped [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Kmers and their gene associations&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		genes [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Genes and their statistics&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		gene_info [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Gene information&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;.tsv&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		test_assoc [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;test_assoc&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		plot_ps [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;plot_ps&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;plot_p.R&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		bonferoni [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;bonferoni&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;bonferoni.R&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		filter [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;filter&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;filter.R&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
	map_kmers [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;map_kmers&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;map_kmers.py&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		summarise_genes [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;summarise_genes&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;summarise_genes.R&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		plot_genes [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;plot_genes&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;black&quot;&gt;plot_genes.R&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];


	} /* }}} */

	subgraph cluster_output_files { /* {{{ */
		fontname=&quot;Sans serif&quot; 
		fontsize=36
		label = &quot;Output files&quot;;
		make_index [ 
			shape=ellipse
			style=filled
			fillcolor=grey
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;make_index&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		tree_plots [ 
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Phylogenetic tree plots&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;png&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		plot_qq [ 
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;QQ-plots&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;html&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		gene_plot [ 
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Gene plots&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;html&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
		web_report [ 
			shape=note
			label=&lt;
				&lt;table border=&quot;0&quot; cellborder=&quot;0&quot; cellspacing=&quot;0&quot;&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;Web report&lt;/td&gt;&lt;/tr&gt;
					&lt;tr&gt;&lt;td align=&quot;center&quot;&gt;&lt;font point-size=&quot;28&quot; color=&quot;blue&quot;&gt;html&lt;/font&gt;&lt;/td&gt;&lt;/tr&gt;
				&lt;/table&gt;
			&gt;
		];
	} /* }}} */

	/* edges {{{ */

	edge [
		arrowsize=3
		penwidth=2
	];

	contigs -&gt; mine_kmers;
	mine_kmers -&gt; kmers;

	contigs -&gt; annotate;
	annotate -&gt; annotations;

	annotations -&gt; pangenome_analysis;
	pangenome_analysis -&gt; phylogenetic_tree;

	phylogenetic_tree -&gt; distance_from_tree;
	distance_from_tree -&gt; distance_matrix;

	phylogenetic_tree -&gt; plot_tree;
	pheno -&gt; plot_tree;
	plot_tree -&gt; tree_plots;

	pheno -&gt; split_phenos;
	split_phenos -&gt; phenos;

	phenos -&gt; pheno_i;
	pheno_i -&gt; test_assoc;
	kmers -&gt; test_assoc;
	distance_matrix -&gt; test_assoc;
	test_assoc -&gt; associated_kmers;

	associated_kmers -&gt; plot_ps;
	plot_ps -&gt; plot_qq;

	associated_kmers -&gt; bonferoni;
	bonferoni -&gt; threshold;
	threshold -&gt; filter;
	associated_kmers -&gt; filter;
	filter -&gt; kmers_filtered;
	
	bonferoni -&gt; stats;

	refs -&gt; ref2bed;
	ref2bed -&gt; refs_bed;

	refs -&gt; bwa_index;
	contigs -&gt; bwa_index;
	bwa_index -&gt; indexes;

	annotations -&gt; annotation2bed;
	annotation2bed -&gt; annotations_bed;

	indexes -&gt; make_ref_list;
	refs_bed -&gt; make_ref_list;
	annotations_bed -&gt; make_ref_list;
	make_ref_list -&gt; map_kmers;

	indexes -&gt; map_kmers;
	kmers_filtered -&gt; map_kmers;
	annotations_bed -&gt; map_kmers;
	refs_bed -&gt; map_kmers;
	map_kmers -&gt; kmers_mapped;
	map_kmers -&gt; stats;
	map_kmers -&gt; gene_info;

	kmers_mapped -&gt; summarise_genes;
	gene_info -&gt; plot_genes;
	summarise_genes -&gt; genes;
	summarise_genes -&gt; stats;

	genes -&gt; plot_genes;
	plot_genes -&gt; gene_plot;

	gene_plot -&gt; make_index;
	plot_qq -&gt; make_index;
	tree_plots -&gt; make_index;
	gene_info -&gt; make_index;
	stats -&gt; make_index;
	make_index -&gt; web_report;

	/* }}} */

}" class="graphviz" /></div>
<span class="target" id="module-bgwas3"></span><dl class="function">
<dt id="bgwas3.assembly">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">assembly</code><span class="sig-paren">(</span><em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.assembly" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Contig assembly</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.mine_kmers">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">mine_kmers</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.mine_kmers" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Kmer mining and counting with fsm-lite</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters</dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>infile</strong> â directory of geneomes (fasta files)</p></li>
<li><p><strong>outfile</strong> â gzipped file of Kmer patterns and genomes they are found in</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="bgwas3.annotate">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">annotate</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em>, <em class="sig-param">idd</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.annotate" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Annotate genomes with prokka (generate gff file for each contig)</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.pangenome_analysis">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">pangenome_analysis</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.pangenome_analysis" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Perform pangenome analalysis with Roary (based on difference in gene
content) and generate a phylogenetic tree (.newick) file for use in mixed</p>
<blockquote>
<div><p>effects association testing</p>
</div></blockquote>
</dd></dl>

<dl class="function">
<dt id="bgwas3.distance_from_tree">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">distance_from_tree</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.distance_from_tree" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Get distances from a phylogenetic tree that has been midpoint
rooted</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.mash">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">mash</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.mash" title="Permalink to this definition">Â¶</a></dt>
<dd></dd></dl>

<dl class="function">
<dt id="bgwas3.plot_trees">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">plot_trees</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.plot_trees" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Plot phylogenetic trees with filters based on clade and phenotype</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.split_phenos">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">split_phenos</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfiles</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.split_phenos" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Split the main tsv file phenotype columns into their own tsv files
(to be used in seperated association tests)</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.plot_phenos">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">plot_phenos</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em>, <em class="sig-param">pheno</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.plot_phenos" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Density plot phenotypes</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.test_assoc">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">test_assoc</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfiles</em>, <em class="sig-param">idd</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.test_assoc" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Association test kmers and phenotypes with pyseer</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.plot_ps">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">plot_ps</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfiles</em>, <em class="sig-param">pheno</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.plot_ps" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Plot p-values (histogram and qqplot) for each association test</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.bonferoni">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">bonferoni</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.bonferoni" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Filter kmers from the output of association based on their p-value,
using bonferoni method to set a threshold</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.ref2bed">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">ref2bed</code><span class="sig-paren">(</span><em class="sig-param">infile</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.ref2bed" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Convert reference gff files into bed files. Also merge overlapping
annotations (eg âCDSâ and âGeneâ entries that are the same, but have
different information. Also remove âRegionâ entries.</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.make_ref_list">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">make_ref_list</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfile</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.make_ref_list" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Make a list of references for kmer mapping (references first, then
draft annotations)</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.map_kmers">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">map_kmers</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfiles</em>, <em class="sig-param">prefix</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.map_kmers" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Map kmers to annotation files</p>
</dd></dl>

<dl class="function">
<dt id="bgwas3.summarise_genes">
<code class="sig-prename descclassname">bgwas3.</code><code class="sig-name descname">summarise_genes</code><span class="sig-paren">(</span><em class="sig-param">infiles</em>, <em class="sig-param">outfiles</em>, <em class="sig-param">prefix</em><span class="sig-paren">)</span><a class="headerlink" href="#bgwas3.summarise_genes" title="Permalink to this definition">Â¶</a></dt>
<dd><p>Count the number of kmer hits per gene, and calculate other
statistics</p>
</dd></dl>

</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">bgwas3</a></h1>








<h3>Navigation</h3>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2019, Gregory Leeman.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.1.2</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/steps.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>