---
output: 
  pdf_document:
    citation_package: biblatex
    number_sections: true
    fig_caption: true
    includes:
      in_header: header.tex
      before_body: cover_page.tex 
bibliography: bgwas3.bib
csl: harvard.csl
---

```{r setup, include=FALSE}
library(dplyr);
library(tidyr);
library(ggplot2);
library(knitr);
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE);
source("tables.R");
```
\fancypagestyle{plain}{%
  \renewcommand{\headrulewidth}{0pt}%
  \fancyhf{}%
  \fancyhead[C]{\footnotesize \thepage}
  \setlength\footskip{0pt}
}
\pagestyle{plain}

# Abstract {-}

Genome-wide association studies (GWAS) are now being applied more often to microbes.
However, bacteria genomes are significantly more variable in both content and sequence than eukaryotic genomes,
meaning these association studies have naturally different, more complex considerations.
These studies utilise short unique DNA patterns, 
referred to as k-mers,
as units of genetic variation 
as opposed to single nucleotide polymorphisms.
Various software solutions are now available 
which can perform the multiple association tests of k-mers,
notable the python based *pyseer*.
However, there is not at present a succinct tool
which pipelines the necessary pre and post data processing and analysis steps
that make an association study whole.

In metabolomics,
thousands of possible measures that can be taken
corresponding to the level of molecules inside or around a cell.
These measures can be considered as traits of that bacteria at a that point in time.
However conducting individual association tests for each metabolomic trait is time and computationally intensive.

I developed a tool, Bgwas3,
which wraps the k-mer association function of *pyseer*
with other open source software tools
into a single installable package.
When run, Bgwas3 takes the simplest of input files
and a configuration file,
and performs gene annotation,
phylogeny estimation,
k-mer association testing,
and k-mer-to-gene mapping  
while also generating automatic visualisations and a web-based report.

The tool, found at https://github.com/g-r-eg/bgwas3, was built with best practices for scientific computing in mind, and is is installable as a conda package.

In this report I discuss it's implementation.

The tool was also used to test the association of 26 traits of *Pseudomonas auriginosa*;
18 relating corresponding to a metabolomic measurement,
and 8 corresponding to either a measure of to antibiotic resistance or bacterial motility.
In 14 of the traits, between 1 and 16450 significant k-mers were found, and were mapped to named genes.

\newpage
\tableofcontents
\newpage

# Abbreviations {-}

CF

: Cystic Fibrosis

DRMAA

: Distributed resource management application

GWAS

: Genome-wide association study

SNP

: Single Nucleotide Polymorphism

TSV

: Tab separated value

LMM

: Linear mixed model

\newpage
\listoffigures
\listoftables
\newpage

# Introduction
## Pangenome wide association studies

Genome wide association studies (GWAS) are are popular tool in human genetics due to their ability identify genetic and phenotype associations in a hypotheses free manner.
Many of the characteristics of microbes suggest they would be ideal candidates for GWAS.
First, their genomes are significantly smaller, which both limits the study size and makes whole genome sequencing easier and cheaper.
Secondly, phenotypes of interest such as virulence and drug resistance tend to be the result of strong selective pressures; meaning they are likely to be controlled by fewer, recent mutations.
There is also good reason to perform association studies with bacteria, as identifying the genetic basis of traits that correspond with pathogenic ability
could lead to better understanding of disease
and lead to new targets for pharmaceutical intervention.

In humans, GWAS have generally use single nucleotide polymorphisms (SNPs) as units of genetic variation.
This is feasible only due to two defining characteristics of humans genomes.
First, the genome experience regular and reliable recombination; meaning only very close loci are in linkage disequilibrium.
This means an SNP that is identified as being significantly associated with a phenotype is likely to be very close to the truly causal polymorphism.
Secondly, the gene sequence and content between humans remains relatively consistent within the species.
This has allowed the development of SNP chip which can interrogate thousands of sites which accurately represent the entire genetic diversity.
It also allows for SNPs to be identified though multiple alignment.
Bacterial genomes possess neither of these qualities.
In bacteria, its very likely that huge regions could be in linkage disequilibrium; obfuscating the true causal loci.
It is also likely that the gene sequence is different, making identifying SNPs through multiple alignment an arduous computational problem that will always be limited to the core genome, ignoring completely genes of the accessory genome.

As an alternative, bacterial GWAS studies are taking a pangenomic approach and some recent studies have instead utilised k-mers: all nucleotide sub-strings of length 'k' found in the genomes.
Initially a concept of genome assembly, k-mers can capture multiple genetic variations such as SNPs, longer deletions/ insertions and recombination sites.
The size of k-mers effects the genetic variation captured, with longer being more specific, but shorter are more sensitive.
Getting k-mers is alignment free, releasing the burden of multiple aliments, but more importantly, allows both the core and accessory genome to be tested for association.

## Metabolomics and the need for multiple phenotype testing

Metabolomics of bacteria involve the large scale study of small molecules within and immediately adjacent to the cell.
Metabolomics can provide a more detailed insight into the underlying biochemical activity and state of a cell than genomics or transcriptomics.
Coupling genomics and metabolomics could allow a better understanding of how complex pathways in the cell have adapted.
The nature of metabolomics means that multiple molecules abundance may be measured and of interest to investigate.

## Scope of work

The foundation of this project was to create a tool, Bgwas3, an integration of mutltiple open source genomics tools and custom scripts written in python, R and bash into a single installable package that facilitates conducting multiple pangenome wide association studies in a single user step.
The tool aims to be unique, accessible and robust; allowing a user to convert only the most basic required input files into valuable and interpretable results including static and interactive visualisations.

# Methods

```{r out.height="95%", fig.align="center", fig.scap="\\label{fig:pipe}Graphical representation of bgwas3", fig.cap="\\label{fig:pipe}Graphical representation of the 18 tasks in the pipeline of bgwas. Grey ellipses encode individual tasks, and are labelled with the task name and then either the external tool used, or the external script which is run where applicable. White nodes indcate files or groups of files, and are labelled with the file format(s). The yellow node indicates the configuration file which may be optionally included as an input to define some of the pipelines paramters. The light blue box highlight steps which are repeated for each phenotype tested, with all other steps only needing to be run once.", out.extra=''}
knitr::include_graphics("_static/pipe.png");
```

## Pipelining with CGAT-Core

I made Bgwas3 with the Pipelining framework CGATCore, [@cribbs_cgat-core:_2019].
CGATCore is extremely portable framework that comes as a dependency free Python package,
but has been used to construct complex pipelines that are scalable to large amounts of data.

With the framework, individual pipelining steps, referred to as 'tasks', are defined as python functions whose arguments include the necessary input files and the expected output files.
With the use of python decorators, the output files of one task can be used as the input arguments of later tasks.
This allows multiple data analysis to be strung together and run in the correct order.
The framework utilises a system of checking the modification date of files, so only runs tasks which have either new or modified input files.
This checking system means a directory of project files may move from one location to another, even between computer, and the framework will still recognise which tasks and steps are out of date.

Another significant feature of CGATCore which makes it stand out amongst other pipelining tools is an interface to control a distributed resource management application (DRMAA) such as PBS-Pro used by Imperial College's resource computing service.
Tasks which process files in parallel can run in parallel as batch processes, and tasks which require either high storage or memory resources it can be distributed to nodes which match these requirements.
This makes executing pipeline such as Bgwas3, which involve multiple computationally intensive intermediary steps faster.

Bgwas3 was written as 18 'tasks' which are visualised in figure \ref{fig:pipe}.
I wrote the pipeline in python, but individual tasks made calls to external tools or scripts I wrote in either python or R.

As input, the tool requires the following three starting files:
1. A directory containing fasta files uniquely named for each bacteria sample.
2. A tab separated value file (TSV) in which the first column matches the names of sample fasta files, and remaining columns corresponding to trait measures. 
3. An optional directory of reference sequences in fasta format accompanied by annotation files in general feature format (GFF).
4. An optional configuration file in .yml format that specifies extra parameters for the running of the pipeline.

## Genome Annotation

In the Bgwas3 task 'annotate' all sample sequences are annotated by the external tool Prokka [@seemann_prokka:_2014]
Prokka first makes predictions on features using a selection of external tools including 
Prodigal [@hyatt_prodigal:_2010] for the annotation of coding region, 
RNAmmer [@lagesen_rnammer:_2007] which identifies for ribosomal RNA genes,
Aragorn [@laslett_aragorn_2004] for transfer RNA genes, 
and SignalP [@armenteros_signalp_2019] for signal leader peptides and Infernal for non-coding RNA.
After feature prediction, the speculative features are queried against a umber of databases including UniProt [@the_uniprot_consortium_uniprot:_2019], RefSeq [@oleary_reference_2016] and Pfam [@el-gebali_pfam_2019].

Draft sequences are annotated in the Bgwas3 pipeline for two reasons:
First, gene annotations are later used to estimate the phylogenetic tree of the samples;
and secondly, significantly associated k-mers are later mapped to the annotated genomes when attempting identifying the k-mer's genetic identity.

## k-mer minig and counting

Bgwas3 integrates the external tool FSM-lite [@valimaki_fsm-lite_2018] to 'mine' and count k-mers.
A benefit of FSM-lite is, unlike other k-mer mining tools such as DSK [@rizk_dsk:_2013], FSM-lite allows a range of k-mer-sizes, as defined by the user, to be mined as apposed to a single length.
Bgwas3 allows the k-mers length to be changed in the pipeline by editing the Bgwas3 configuration yml file values 'fsm_k-mer_min' and 'fsm_k-mer_max'.

## Phylogeny estimation and covariate generation

Bgwas3 currently takes a pangenomic approach to phylogeny estimation, as in the relative presence and absence of genes are used to calculate the distance between samples.
A phylogenetic tree is estimated with the tool Roary [@page_roary:_2015].
In summary, Roary determines genes which fall within the core genome, and then performs clustering of isolates based in the constitution of the variables accessory genome.
One of Roary's outputs, a tree in the common newick format, which is then converted into a distance matrix TSV file.

The single reasoning for predicting phylogeny in the Bgwas3 pipeline is to use the distances defined in the distance matrix as covariates in the later association testing.

## k-mer association testing

The typical analytical strategy implemented in GWAS is some form of linear regression.
Standard regression assumes that data is are identically and independentlly distributed.
However, due to population stucture, this is almost always an incorrect assumption in GWAS studies,
and if that assumption is made incorrectly, 
other genetic polymorphisms will be incorecctly identified as associated to the trait of interest.

Most GWAS techniques involve implement a statistical method to to control for the confounding population structure potential.
In human studies, for example, its not abnormal to remove highly related individuals, though this will ultimately reduce the power of the study by decreasing sample size.
Another popular techique involves peforming multiple dimensional scaling, and then including significant principal compoents as fixed effects in the linear model.
Bacteria experience a much stronger population structure as a result of clonal reproduction, and so a need for control becomes more important.

Bgwas3 implements a linear mixed model (LMM) as provided by pyseer to tackle population,
specifically factored spectrally transformed LMM [@lippert_fast_2011].
LMMs tackle confounders by using measures of genetic similarity as random effects within the linear model.
Given a good measure of hierarchical relatedness between samples, the mixed model is generally preferred, and has been shown in past studies to control the inflation of p-values better [@lees_genome-wide_2017].

Bgwas3 uses the distance matrix from the estimated phylogeny as covariates.
In the Bgwas3 task 'test_assoc' an association test for each phenotype is run using all the k-mers.

The output of 'test_assoc', a list of all k-mers and statistics relating to their association to the given phenotype, are then filtered by their P-value though bonferoni correction.
The Bonferroni correction is a multiple-comparison correction used when several statistical tests are performed, to limit the number of false-positive results. 
It's vital in GWAS studies to apply some filtering, as the number of independent tests are so high that it is very likely that spurious rare events (like chance false positive) will occur.
Only significant k-mers are then used in later steps of the pipeline.

## k-mer mapping

```{r,, fig.cap="\\label{fig:map_k-mers}Graphical representation of algorithm to map k-mers to reference genes"}
knitr::include_graphics("_static/map_kmers.png");
```

The Burrows-Wheeler Alignment Tool (BWA) [@li_aligning_2013] is primarily used to map significant k-mers to genes.
The tool first requires that sequences are converted into an FM-index: a data-structure similar to a suffix array. This is performed with the Bgwas3 task 'bwa_index' (figure \ref{fig:pipe}) on all sample and reference fast files.

For use with bedtools [@quinlan_bedtools:_2010], all annotation files in GFF format are converted into BED format, and subsequently filtered with the tasks 'annatation2bed' and 'ref2bed' repsectively.
Sample annotations from Prokka are filtered to include only those annotation which correspond to a named gene, and so exclude ones which relate to hypothetical proteins or non-genes.
Entries in the reference annotation file which correspond to the same loci are merged, collating the information from both.

The task 'map_k-mers', executes a python script I wrote whose algorithm is visualised in figure \ref{fig:map_k-mers}.

In summary, the algorithm works as follows:

1. Generate an artificial multi-fasta file with an entry for each unmapped k-mer.
2. Choose the next reference file.
3. Attempt to align the artificial fasta file to the reference fasta file with BWA-mem.
4. Generate an artificial BED file where each entry corresponds to a successful alignment with BWA-mem. 
5. Use 'bedtools intersect' command line tool to compare and retrieve matches in the query and reference BED files.
6. Harvest all information about each intersect stored in a gene info file, 
8. Mark the k-mer as mapped.
7. Repeat until all k-mers mapped or until all references have been used

## Output and Visualisation

Bgwas3 automatically generates multiple figures which are then integrated into a final web based report.
Static visualisations are made with external scripts I wrote in R that make use of ggplot2 package [@wickham_ggplot2:_2016].
The task 'plot_ps' generated a quantile-quantile plot from all unfiltered p-values (see figure \ref{fig:genes}), and a phylogenetic tree visualisation is built from the newick file from Roary, and the input phenotype file utiliseing external R package ggtree [@yu_ggtree:_2017]. (see figure \ref{fig:tree}).

Finally, for each phenotype, a plot of genes is made in which the following features are visually encoded (see figure \ref{fig:genes}):

- Maximum $-log_{10}\left ( Pvalue \right )$ of k-mers mapped to that gene
- Average beta value (effect size)
- Average allele frequency
- Number of k-mers 'hits;

Finally, a web based report is generated with the task 'make_index'. The report includes an interactive table of all phenotypes that can be sorted by the number of significant k-mers found and the number of genes., such as the protein product.
Links then take the user to an interactive version of the gene plot (see figure \ref{fig:genes}) in which all information that was originally stored in the annotation files can be retrieved, such as the protein product.

# Results with test data

**Dataset**
For the development, testing, and evaluation of Bgwas3 a dataset corresponding to genomes and phenotype measurements of *Pseudomonas aeruginosa* was used.

Cystic fibrosis (CF) is an autosomal recessive disorder that, due to a single gene mutation,
causes a patient to have defective trans-membrane regulator protein.
This protein is situated in epithelial cells that make up the mucus membranes of the body, and is primarily responsible for transporting chloride ions and bicarbonate [@riordan_identification_1989].
The dysfunctional form of this protein ultimately limits the osmotic movement of water,
and the mucus at these membranes remain viscous and immotile.
In the lungs of healthy individuals, a less viscous mucus is able to be transported by cilia out of the lungs,
but in patients with CF, the stagnant mucus instead becomes an ideal environment for bacteria to propagate,
and so patients experience at first episodic, but then chronic infection of the lungs.
With prolonged antibiotic use, the fast generation time of bacteria mean resistant strains soon develop, and the prologned inflammation leads to respiratory failure and death.

Pseaudomonas auruginosa is a gram negative bacteria which can easily integrate exogenous DNA into its own genome, making it able to adapt to antibiotic pressures rapidly.
For this reason, it is a common hosptial-aquired infection [@quick_seeking_2014].
It is also one of the primary bacteria present in the lungs of late stage and terminal patients with cystic fibrosis when the lungs function starts to decrease.
For this reason, undesrstanding the genetic adaptations pseudomonas experience when in chronic state is significantly important in talking this disease

In a previous study [@behrends_metabolic_2013] 91 strains were collected over a period of 24 years from 18 patients suffering from Cystic Fribrosis.
Assays of antibiotic resistance, bacterial motility, and later metabolomic measurements were taken.

Bgwas3 was used to test the genetic association of 26 traits (see tables \ref{tab:non_met} and \ref{tab:met}).
18 traits correspond to metabolomic measurement
and 8 corresponding to either a measure of to antibiotic resistance or bacterial motility.

```{r results='asis'}
dat_non_met %>% knitr::kable(caption = "\\label{tab:non_met}Table of non-metabolite phenotypes" );
```

```{r results='asis'}
dat_met %>% knitr::kable(caption = "\\label{tab:met}Table of metabolite phenotypes");
```

**Phenotype data preprocessing**

```{r, out.height="90%", fig.align="center", fig.cap="\\label{fig:dens}Density plots"}
knitr::include_graphics("_static/dens.png");
```

When association testing a qualitative variable with linear regression, it is generally assumed that the variable follows a normal distribution.
When this assumption proves not to be true, and the continuous trait displays severe skewness, the regression may fail to control for type-1 errors (false positives).
A popular statistical technique in GWAS involes transforming the data into a normal form.
Often, a simple log transformation can be sufficient.
Recently, rank based inverse normal transformations (INT) have become popular among genetics researches [@mccaw_operating_2019].

Prior to running Bgwas3, the 26 phenotypes were separately log transformed and transformed using INT, essentially tripling the number of phenotypes tested to 78.
Density plots of the unadjusted and transformed traits are visualised in figure \ref{fig:dens}.

**Phylogeny estimation**

All 91 genomes where annotated and 14643 unique genes were identified with Prokka.
As identified by Roary, these genes were found in >99% of the genomes, and constitute the core genome, leaving 10005 in the accessory.

From these genes, a phylogenetic tree was estimated, and visualised (figure \ref{fig:tree}).

\newpage
\blandscape
```{r, out.height="100%", fig.align="center", fig.cap="\\label{fig:tree}Density plots"}
knitr::include_graphics("_static/tree.png");
```
\elandscape
\newpage

An inspection of the tree shows that, in general, samples from the same patient are generally clustered together, which leads me to believe the phylogeny estimate is somewhat accurate.
It is not unexpected that the dividing between patients is not perfect, due to possible cross contamination and horizontal gene transfer, as many of the patients originated from the same hospital.
Traits of antibiotic resistance seem to be best associated with phylogeny, seen as red clusters in figure \ref{fig:tree}.

**k-mer mining**

497827 unique k-mers were mined of sizes between 9 and 100 base pairs in length. 

**Association results**

```{r echo=FALSE, error=FALSE, message=FALSE, results='asis', fig.cap="\\label{table:genes}Table of metabolite phenotypes"}
dat_genes %>% knitr::kable();
```

```{r, out.height="100%", fig.align="center", fig.cap="\\label{fig:genes}Genes"}
knitr::include_graphics("all.png");
```

13 of the 26 traits, either in their untransformed state or log transformed state had significant k-mers that passed the bonferoni threshold of p < 1.004*10-7, and were mapped to one or more genes.
The trait with the most number of signifiant k-mers was the one corresponging to Leucine, in which 16450 significant k-mers were identified and mapped to 338 genes.
By inspecting the QQ-plots, it becomes apparent that the statudyhas mixed results in regard to accounting for population structure. A well controlled study would present a smooth QQ plot wtith little or no large ridges. However, judging from the QQ-plots, most of the studues were well controlled for low p-values.

# Discussion

A significant goal of bgwas3 was to implement a useful and reusable tool that may be used outside the scope of this project.
As such, Bgwas3 was built with the best scientific computing practices in mind to aid reproducability and ease of use.
Bgwas3 is packaged and is currently installable as a conda recipe (https://anaconda.org/g-r-eg/bgwas3) whereby all external dependecies are included, or through pypi (https://pypi.org/project/bgwas3/).
It also has some testing using continuous integration (https://travis-ci) and has been success2ully deployed on OSX and Ubuntu.
The project resides on a github repository, and is under a GNU General Public License v3.0, and external additions and development is encouraged using the standard git protocols.

## Future work

Through the development of Bgwas3 it's use on test data, many limitations and possible areas for addition became apparent.
However, because Bgwas3 was written using a pipelining framework, it is possible to add additional methods of individual steps and add more post analysis tasks without compromising what has already been built.
It is hoped that development of bgwas3 will continue, and some of these new additions will be included.

**Phylogeny estmation**
Currently, Bgwas3 implements only a pangenomic approach approach of distance estimation.
This approach, gene presence and absence has not been critically evalulated or compared with other methods, though many other possbible methods exist.
For example, a common approach to estimating phylogeny involves peforming a multiple alignment of the core geneome, and using the difference in SNP content as distance estimates.
Tools such the general multiple aligner Mummer4 [@marcais_mummer4:_2018] and the bacteria-focused tools ParSnp [@treangen_rapid_2014] and ClonalML [@didelot_clonalframeml:_2015] suite attempt this mutliple alignement, while peforming inference of recombination.


**Interpretation of k-mer results**
K-mer-based GWAS is relatively new, but already some tools are finding new and possibly better ways to interpret significant K-mers in a manner besides mapping them to genes.
This is the primary feature of bacteria GWAS tool DBGWAS [@jaillard_fast_2018].
DBGWAS uses all the k-mers from the study to rebuild a de-bruijn graph in a manner similar to genome assembly. 
After k-mers are then tested for assoctiion, the associated kmers are then visualised as sub-graph of the full de-bruijn graph, where shape and colour encode the statititics of significance and p-value.
This allows clusters of genes which are both close to one another and significicantly associated to be quickly identified.
Another possible addition to Bgwas3 which would significantly advance its usefulness as a tool is pathway enrichment analysis.
For example, in paper by @marvig_convergent_2015 a sample of pseudomonas aeruginosa were analysed using a clone based methods and identified the convergent eveolution of 52 genes.
As a final analysis of the genes, they were grouped together according to their function as defined by PseudoCap, and through enrichment analysis, ovr-represented classes were identfifed.

**Control for population structure**
Other software packages exist for performing GWAS experiments with bacteria that don't rely on the LMM approach of population control. 
For example, the programme as treeWas [@collins_phylogenetic_2018] instead generates a simulation of a null genetic dataset using parameters of the emperial dataset phylogeney. 
From the simulated dataset it can then peform association tesing, and build a null distrubution of score statistics under the null hypotheses of no association. 
This provides a strict control over the false popsitveity rate.

# References
