---
header-includes:
  - \usepackage{setspace}\doublespacing
  - \usepackage[nottoc]{tocbibind}
output: 
  pdf_document:
    number_sections: true
    fig_caption: true
    includes:
      before_body: cover_page.tex 
bibliography: bgwas3.bib
---

```{r setup, include=FALSE}
library(dplyr);
library(tidyr);
library(ggplot2);
library(knitr);
knitr::opts_chunk$set(echo=FALSE, message=FALSE, warning=FALSE);
source("tables.R");
```

# Acknowledgements {-}

I would like to thank my supervisor Antonio Berlanga for giving me the opportunity to undertake this project in which I learned to much and gained so many valuable skills for which I am sure will value in my continuing career.

\newpage

# Abstract {-}

Genome-wide association studies (GWAS) are now applied more often to microbes.
However, bacteria genomes are significantly more variable in both content and sequence than eukaryotic genomes,
meaning these association studies have naturally different, more complex considerations.
These studies utilise short unique DNA patterns, 
referred to as Kmers,
as units of genetic variation 
as apposed to single nucleotide polymorphisms.
Various software solutions are now available 
which can pefoerm the multiple assocation tests of kmers,
notable the python based *pyseer*.
However, there is not at present a succint tool
which pipelines the necessary pre and post data processing and analysis steps
such as associating the Kmers to genes,
that make an association study whole.

In metabolomic studies of bacteria
there are thousands of possible measures that can be made
which can be considered as traits of that bacteria at a that point in time.
Coupling genomics and metabolomics may provide a powerful technque to understanding how a bacteria has adapted,
and in the scope of disease,
may aid in the understanding of the function of drug resistance and other virulence features.
However conducting indivudal assocation tests for each trait may prove time and computationally intesnsve.

I developed a tool, *bgwas2*,
which wraps the Kmer association function of *pyseer*
with other open source software tools
into a single installable package.
When run, *bgwas3* takes the simplest of input files
and a configuration file,
and peforms gene annotation, phylogeny estimation, kmer association testing, kmer-gene mapping and finally visualisation with a automatically generated web report.

In creating the tool, best practices for computational biology where excercised; resulting in a final tool that expresses approprate testing, documentation and packaging; meaning it can easily be utilised by other in the future.

The tool was also used to test the assocation of 26 traits 
- 18 relatiing correspjoning to a metabolomic measurement,
and 8 corresponding to either a measure of to antibiotic resistance or bacterial motility.
In 14 of the traits, between 1 and 16450 significant Kmers were found, and were mapped to Genes.

\newpage
\tableofcontents
\newpage

# Abbreviations {-}

GWAS

: Genome-wide assocation study

DRMAA

: Distributed resource management application

TSV

: Tab seperated value

\newpage
\listoffigures
\listoftables
\newpage

# Introduction
## Pangenome wide association studies

Genome wide association studies have been successfuly used to identify genetetic and phenotype assocations in a hypthoses free manner.
Characteristics of microbes suggest they would be ideasl candidates for GWAS:
their genomes are small which limits the study size; they exude much less phenotypic flexibility; and, generally, phenotypes that are of interest such as virulence ans drug resisistance tend to be the result of strong selective pressures.
There is also good reason to peform assocation studies with bacteria.
as identifying the genetic basis of traits that correspond with its pathogenetic ability
could lead to better understanding of disease
an lead to new targets for pharmaceutical intervention

Commonly, GWAS have focued on single nucelotide polymorphisms (SNPs) as units of genetic variation.
SNPs are useful in organsims where recombination is reliable and genomes are relatively consistant among populations.
Because of linkage disequilibrium, identigying a highly associated SNP often means that the true associated loci is nearby

However, the nature of bacteria replication makes the use of SNPs in association studies less useful.
For one, significat regions of the geneome can be in linkage disequilibrium;
and though some species experience high rates of recombination, the recombination is not as reliable or consistent as in eukaryotes to reliably reduce linkage.

Secondly, bacteria experince much higher trans? during recombination. 
Therfore, bacteria of the same species? can vary consideriably in both order of gene content, and in the accessory genome - regions of the geneome which differ.

Therfore, first finding SNPs becomes a tricky process which involves multiple alignments. Such tools such as mummer (and other?) do meet this problem, but even still, SNPs assocated with ony the core genome can be retrieved
and does not consider the huge variable accessory genome in bacteria/ will not encapsulate the ful0

As an alternative, some recent studies have instead utlised K-mers.

Kmers are simply variable length sequences of DNA
Initially a concept of genome assembly
Can caputre multiple geneetic variations 
SNPs, 
longer deletions/ insertions 
and recombination 
The size of kmers effects the genetic variation captured
Longer more specificS
Shorter are more sensitive

Getting kmers is alignment freee, releasignt he burden of ultie alighnemt
But more specifically utilisng kmers means the core ans accesory geneome can be tested.

*Seer* [@lees_sequence_2016] and its python reimplementation *pyseer* [@lees_pyseer:_2018]
dbgwas

## The Need for Multiple Assocation Tests - Metabolomics

Metabolomics provides a detailed snapshot of an organismâ€™s physiological state through the quantification of hundreds to thousands of small molecules (Zampieri et al., 2017). 

Coupling genomics and metabolomics to study Pseudomonas aeruginosa in unique clinical samples can allow us to understand how bacteria adapt to the lung environment.

The nature of metabolimics measn that multiple molecules abundance may be measured and of interest to investigate.

Determint the genetic basis of metabolomics of P ing the context of chronic invextions

## Scope of work

The foundation of this project was to create a tool, *bgwas3*,
an integration of mutltiple open source genomics tools 
and custom scripts written in python, R and bash
into a single installable package 
that facilitates conducting multiple pangenome wide association studies
in a single user step.

The tool aims to be comprehensive and robust enough to convert only the most basic required input files 
into valuable and interpretable results 
including static and interactive visualisations.

When run on a node within a computer cluster, 
tasks that are computationaly intensive or long 
are distrubuted among the cluster,
and can run simultaneuously.

Though the additional confugiureation file, the specifics of multile intermiedary steps can be altered, which can have variable effects on the final results.

In light of the necessity to peform multpple.. (\ref{fig:pipe})

# Implementation 

```{r out.height="90%", fig.align="center", fig.cap="\\label{fig:pipe}Pipeline of bgwas"}
knitr::include_graphics("_static/pipe.png");
```

## Pipelining with CGAT-Core

I made *bgwas3* with the piplining framework CGATCore, [@cribbs_cgat-core:_2019].
CGATCore is extremeley portable framework that comes as a dependency free Python package,
but has been used to construct complex pipelines that are scalable to large amounts of data.

With the framework, individual pipelining steps, 
referred to as 'tasks',
are defined as python functions 
whose arguments include the necessary input files 
and the expected output files.
With the use of python decoraters, the output files of one task can be used as the input arguments of later tasks.
This allows multiple data anlysis to be strung together and run in the correct order.
The framework utilises a system of checking the modifcation date of files,
so only runs tasks in which their are either new input files, or input files modified.
This checking system means a directory of project files may move from one location to another, even between computer,s
and the framework will still recognise which tasks and steps are out of date.

Another significant feature of CGATCore which makes it stand out amongst other pipelining tools is an interface to control a distributed resource management application API (DRMAA) such as PBS-Pro used by Imperial College Londons resource computing service.
Tasks which process files in parallel are large can run in parallel as batch processes,
and tasks which require either high storage or memory resources it can be distributed to nodes which match these requirements.
This makes executing pipeline such as *bgwas3*, which involve multiple computationally intesive intermediary steps faster.

*bgwas3* was written as 18 'tasks' which are visualised in figure \ref{fig:pipe}.

I wrote the pipeline in python, but individual tasks made calls to external tools or scripts I wrote in either pyton, R and bashscript.

As input, the tool requires the following three starting files:

1. A directory named 'contigs' containing fasta files uniquely named for each bacteria sample
2. A tab seperated value file (TSV) in which the first column matches the names of sample fasta files, and remaining columns correspong to phenotype values.
3. An optional directory of reference sequences in fasta format accompanied by annotation files in gff format.

## Genome Annotation

In the *bgwas* task 'annotatate' all sample sequences are annotated by the external tool Prokka [@seemann_prokka:_2014]
Prokka first makes predictions on features using a selection of external tools including 
Prodigal [@hyatt_prodigal:_2010] for the anotation of coding region, 
RNAmmer [@lagesen_rnammer:_2007] which identifies for ribosomal RNA genes,
Aragorn [@laslett_aragorn_2004] for transfer RNA genes, 
and SignalP [@armenteros_signalp_2019] for signal leader peptides and Infernal for non-coding RNA.
After feature prediction, the speculative features are queried against a umber of databases including UniProt [@], Refseq [@] and Pfam.

Draft sequences are annotated in the *bgwas3* pipeline for two reasons.
First, gene annotations are later used to estimate the phylogenetic tree of the sample,
and secondly, significantly associated Kmers are later mapped to the annotated genomes when attempting identifying the Kmer's genetic identity.

## Kmer minig and counting

*bgwas3* integrates the external tool FSM-lite [@valimaki_fsm-lite_2018] to 'mine' and count Kmers.
A benefit of FSM-lite is, unlike other Kmer mining tools such as DSK [@rizk_dsk:_2013], FSM-lite allows a range of kmer-sizes, as defined by the user, to be mined as apposed to a single length.
*bgwas3* allows the Kmers length to be changed in the pipeline by editing the configuration yml file values 'fsm_kmer_min' and 'fsm_kmer_max'.

## Phylogeny estimation and covariate generation

*bgwas3* currently takes a pangenomic approach to phylogeny estimation;
as in the relative prescence and absence of genes are the considering factor which the distance between samples are determined.
An phylogenetic tree is estimated with the tool Roary [@page_roary:_2015].
In summary, Roary determnes genes which fall within the coregeneome,
and then peforms clustering of isolates based in the constetution of the variable accessory geneome.
One of roary's outputs, a tree in the common newick format,
which is then converted into a distance matrix tsv file using a python script.

The single reasoning for prediciting phylogeny in the *bgwas3* pipeine is to use the distances defined in the distance matrix as covariates in the later assocation testing.


## Kmer association testing

The typical analytical strategy GWAS involves some form of linear regression.

It is almost always wrong to assume that a trait is the only.

If a phenotype is linked with ancestry, thon other geneetic polymorphsm may be incorrextly identified as linked to the trait of interest.

Many methods are deployed in an attempt to reduce the inevistable effect of population structure has on associaion studies. 

In human studies, for example, its not abnormal to remove highly related individuals, though this will ultimately reduce the power of the study by decreasing sample size.

Another popular techique involves peformng some multiple dimensianal scaling, and then including significant principal compoents as fixed effects in the model.


*bgwas* instead attempts to tackle the problem of population structure by one of the three available tests of assocation availabe in the pyseer [@lees_pyseer:_2018] suite:
linear mixed model (LMM),
specifically factored spectrally transformedas LMM [@lippert_fast_2011].

LMMs can capture all of these confounders simultaneously, without knowledge of which are present and without the need to tease them apar

Roughly speaking, LMMs tackle confounders by using mea sures of genetic similarity to capture the probabilities that pairs of individuals have causative alleles in common. Such measures include those based on identity by descent 10

Given a means of determining the hierarchichal relatedness between samples, the mixed model is generally preffered, and has been shown in past studies to control the inflation of p-values better [@lees_genome-wide_2017].

*bgwas* uses the distance matrix from the estimated phylogeny as covariates.

In the *bgwas* task 'test_assoc' and association test for each phenotype on all the Kmers. In a computer cluster, these individual tests may be run simultaneuously.

## Bonferoni correction

The output of pyseer, a list of all kmers and statistics relating to their association to the given phenotype, are then filtered by their p-value though bonferoni correction.

This is done in two steps. First the bonferoni threshold is calculated from the number of kmers minded.

Bgwas3 allows the user to change the alpha level

The Bonferroni correction is a multiple-comparison correction used when several dependent or independent statistical tests, utilised to limit the number of spurious positive tests.
f multiple hypotheses are tested, the chance of a rare event increases, and therefore, the likelihood of incorrectly rejecting a null hypothesis (i.e., making a Type I error) increases

Only significant kmers are then used in later steps of the pipeline.

## Significant Kmer mapping

```{r,, fig.cap="\\label{fig:make_kmers}Graphical representation of algorithm to map Kmers to reference genes"}
knitr::include_graphics("_static/map_kmers.png");
```

The Burrows-Wheeler Alignment Tool (BWA) is primarily used to map significant Kmers to genes.

The tool first required that sequences are first converted into an FM-index, an data-structure similar to a suffix array. This is peformed with the *bgwas* task 'bwa_index', on all sample and reference fast files.

For use with bedtools, all annotation files in gff are similaryl converted into bed format, and then subsequently filtered with the tasks 'annatation2bed' and 'ref2bed' repsectively.
Sample annotation from prokka are filtered to include only those annotation which correspond to a named gene, and so exclude ones which relate to hypothetical proteins or non-gene loci.
Reference annotations which correspond to the same region but are present as two sperate entries are merged, keeping the information from both, as this was found to be a common nomenclature in well annotated references. 

The task 'map_kmers', executes a python script I wrote whose algorithm is visualised in \ref{map_kmers}.

In summuary, the algorithm works as follows:
- A fake multi-fasta file is generated with an entry for each unmapped, significant kmer
- The next reference is chosen
- The command-line tool 'bwa-mem [@li_aligning_2013] attempts to align to the reference
- A fake bed file is generated with an entry correspoinding to each successful alignement with bwa mem
- The query bed file is compared to the reference bed file using the command line tool 'bed tools intersect'
- The information about each intersection is harvested and stored in a gene info file, while the kmer is marked as mapped
  Repeat until all kmers mapped or until all references have been used

## Visualisation

An important feature of *bgwas3* is the automated generation of multiple figures, integrated into a final wbe based report.

Static visualisations are made with external scripts I wrote in R that make use of ggplot2 package [@wikham_ggplot2:_2016].

The task 'plot_ps' generated a quantile-quantile plot from all p-values unfilterd. This may give
See fig

A single phylogenetic tree is built from the newick file and phenotype file utiliseing external R package [@ggtree]. See figs

Finally, for each phenotype, a plot of genes is made in which the following features are visually encoded

- Maximum -log10(p-value) of Kmers mapped to that gene
- Average beta value (effect size)
- Average allele frequency
- Number of kmers 'hits'

Finally, a web based report which incorporates all the static visualisations, and remakes an interactive visualisation anaologous tot he gene plot, is buil with the task "make_index".

A datatable which can be filtered by number of genes

The interactive visualisation for each gene
Shows

## Scienitif computing practices

A significant goal of bgwas3 was to implement a useful and resuable tool that may be used outside the scope of this singlular project.

As such, various software development principles and concepts were applied to the project

CGAT-core is implemented in Python 3 and installable via Conda and PyPI with minimal dependencies. We have successfully deployed and tested the code on OSX, Red Hat and Ubuntu. We have made CGAT-core and associated repositories open-source under the MIT licence, allowing full and free use for both commercial and non-commercial purposes. Our software is fully documented (https://pypi.org), version controlled and has extensive testing using continuous integration (https://travis-ci. org/cgat-developers.) We welcome community participation in code development and issue reporting through GitHub


### Testing with PyTest and Travis

Pytest and 

### Documentation with Sphinx

### Packaging with Conda
# Results

**Dataset**
For the development, testing, and evaluation of *bgwas3* a dataset of genomes and phenotypes related to *Pseudomonas aeruginosa* was used.


Cystic fibrosiis is a an autosomal recessive disorder that, due to a single gene mutation, leads to a defection trans-membran regulator protien. This protein is situated in epithelilial cells that make up the mucus membranes of the body, and is primarily responsible for transporting chloride ions and bicarbonate to these membranes. 
The dysfunction of this protein ultimately limits the osmotic movement of water, and these membranes remain viscous and immotile.
In the lungs of healthy individuals a less viscous mucus is able to be transported by cilia out of the lungs. 
In patients with CF, the mucus instead becomes an ideal environment for bacteria to propogate, and so CF pateints experience at first episodic, but then chronic infection of the lungs.
Even with prolonged antibitic use, the fast generation time of bacteria mean resistant strains soon debvelop, and the prologned inflammation leads to respiratory failure and death.

Pseaudomonas euruginosa is a gram negative bacteria which can easily integrate exoenous DNA into its own genome, making it able to adapt to antibiotic pressures rapidly. 
Dor this reason, it is a common hosptial-aquired infection.
It is also one of the prmary bacteria present in the lungs of late stage and terminal patients with cystic fibrosis when the lungs function starts to decrease.
For this reason, undesrstanding the genetic adaptations pseudomonas experience when in chrnic state is significantly important in talking this disease

In a previous study [@behrends_metabolic_2013] 91 strains were collected over a period of 24 years from 18 patients suffering from Cystic Fribrosis.

*bgwas3* was used to test the association genetic association of 26 traits (see tables \ref{tab:non_met} and \ref{tab:met})
18 relatiing correspjoning to a metabolomic measurement
and 8 corresponding to either a measure of to antibiotic resistance or bacterial motility.


```{r results='asis'}
dat_non_met %>% knitr::kable(caption = "\\label{tab:non_met}Table of non-metabolite phenotypes" );
```

```{r echo=FALSE, error=FALSE, message=FALSE, results='asis'}
dat_met %>% knitr::kable(caption = "\\label{tab:met}Table of metabolite phenotypes");
```

```{r, fig.cap="\\label{fig:cormat}Correlation matrix of phenotypes"}
source("plot_cormat.R");
plot_cormat;
```

From the corrlation matrix, it appears that most phentyes corresponding to antibiotic resistance tseem to be somewhat positive correlated, but similarly negatively corelatedto measres of motility.
Stone levels of positive correlation were foudn between metabolites valine, with isoleucine and isoleucine.

## Phenotype data preprocessing 

```{r, out.height="90%", fig.align="center", fig.cap="\\label{fig:dens}Density plots"}
knitr::include_graphics("_static/dens.png");
```

When assocation testing a qualitatitve variable with linear regression, it is generally assumed that the variable follows a normal distrubution.
When this assumption proves not to be true, and the continious trait displays severe skewness the regression may fail to control for type-1 errors (ref)

When a continous trait is non-normal,
a popular statistical technique in GWAS involes transforming the data.
Often, a simple log transformation can be sufficient.
Recently, rank based inverse normal transformations (INT) have become popular among genetics researches.

Prior to running *bgwas*, the 26 phenotypes were seprately log transformed and INT, essesintially trippling the number of phenotypes tested to 78. 
Density plots of the unadjusted and transformed traits are visualised in figure \ref{fig:dens}.

## Phylogeny estimation

All 91 geneomes where annotated and 14643 unique genes were identified with Prokka.
As identified by Roary, these genes were found in >99% of the genomes, and consitute the core genome, leaving 10005 in the accessory.

From these genes, a phylogenetic tree was estimated, and visualised (figure \ref{fig:tree}).

```{r, out.height="90%", fig.align="center", fig.cap="\\label{fig:tree}Density plots"}
knitr::include_graphics("_static/tree.png");
```

The patient of which samples were sourced from are visually encoded by the color of the tooltips.
An inspection of the tree shows that, in general, samples from the same patient are generally clustered together, which leads me to believe the phylegenic estimate is not completely untrue.
It is not unexpected that the deviding between patients is not perfect, due to possible transfection.

Traits of antibiotic resistance seem to be best associated with phylogeny, as seen in fig.

## Kmer mining
497827 unique kmers were mined of sizes between 9 and 100 base pairs in length. 

## Association results
14 of the 26 traits had significant Kmers that passed the bonferoni threshold of p < 1.004*10-7
Of the 26 traits 14 of the traits, between 1 and 16450 significant Kmers were found, and were mapped to Genes (see \ref{tab:genes}).
The trait with the most number of signifiant kmers dound was the one corresponging to Leucine, in which


```{r, out.height="90%", fig.align="center", fig.cap="\\label{fig:genes}Gens"}
knitr::include_graphics("_static/genes.png");
```

```{r echo=FALSE, error=FALSE, message=FALSE, results='asis', fig.cap="\\label{table:genes}Table of metabolite phenotypes"}
dat_genes %>% knitr::kable();
```


# Discussion
In this work...
## Genome annotation

## Kmer mining

The tool fsm

Prokka is a good tool
?Settings

## Phylogeny prediction

Currently, *bgwas3* implements only a pangenomic approach approach of distance estimation.
There are other tools which involve alignment of the core genome and snps...
May or may not be bettter 
Reintroduce the problem of a large multiple alignment.




The strong LD caused by the clonal reproduction of bacterial populations means that non-causal k-mers may also appear to be associated.?



# References
